src/Parsers/ASTRolesOrUsersSet.h

```diff
@@ -25,6 +25,8 @@ class ASTRolesOrUsersSet : public IAST
     bool id_mode = false;         /// whether this set keep UUIDs instead of names
     bool use_keyword_any = false; /// whether the keyword ANY should be used instead of the keyword ALL
 
+    bool none_role_parsed = false; /// whether keyword NONE has been parsed
```

Comment:
Why do we need this extra variable? If NONE has been parsed and it's the only thing that has been parsed then `empty()` returns true and so this variable is not necessary to check this case.

Reply:
This is used to check this usecase and throw an error to warn user it a wrong sql:  
`GRANT NONE, role1 TO user1 WITH REPLACE OPTION;`

Reply:
`GRANT <list_of_roles> WITH REPLACE OPTION` means to revoke all the currently granted roles and then grant all the roles specified in `<list_of_roles>`. If `<list of_roles>` contains only `NONE` it should not grant anything; and if it contains `role1, role2, NONE, role3` it should grant `role1`, `role2`, and `role3`. I mean, yes, there is usually no reason to add `NONE` to `<list of_roles>`, however I don't think it's correct to prohibit that.

Reply:
`GRANT NONE WITH REPLACE OPTION` is not a some kind of exception from how this commands works in the most common case. `GRANT <list_of_roles> WITH REPLACE OPTION` means to revoke all the currently granted roles and then grant all the roles specified in `<list_of_roles>`. Therefore `GRANT NONE WITH REPLACE OPTION` should revoke all the currently granted roles and then grant `NONE` (grant NONE == don't grant anything).

Reply:
Sorry, I missed your reply.  Sounds reasonable, I'll submit in a new pr, thanks you.